core/src/entry.rs

```diff
@@ -299,6 +287,20 @@ impl EntrySlice for [Entry] {
         );
         res
     }
+
+    fn validate_hash_separation(&self, hashes_per_tick: u64) -> bool {
```

Comment:
I'm not a fan of the name of this function: `validate_hash_separation()`

`validate_hash_count()`?

Reply:
Sure, do you think that it would be surprising that `validate_hash_count` rejects an entry slice that doesn't end in a tick entry? I want to enforce that but maybe it would be better do in another place

Reply:
I don't think we want to do that.  There's no guarantee an entry slice always ends with a tick

Reply:
Ok, what about slots? Should we enforce that they end in ticks?

Reply:
I don't think we need that as the number of ticks in a slot is something that's computed dynamically by the bank on each node.  The incoming ledger data itself has no "slot marker".
eg, https://github.com/solana-labs/solana/blob/5468be2ef971b6f4d66af18a70036a3fb975e20e/runtime/src/bank.rs#L807-L810

Reply:
@mvines do you think it's worth fixing https://github.com/solana-labs/solana/blob/5468be2ef971b6f4d66af18a70036a3fb975e20e/runtime/src/bank.rs#L789 while I'm in the area? I don't think we can say that hashes per tick is enforced throughout the blockchain until that is fixed. I can try to tackle it in a follow up PR

Reply:
Yeah for sure.  It's time to assert there and see what breaks.  Ideally nothing else 😱

Reply:
@mvines I think we need to enforce that blocks end in ticks. Otherwise a bank will reach `max_tick_height` (`slot * ticks_per_slot`) and then can still append more hashes in the form of tx entries.

Reply:
I added a check for whether blocks end in ticks. It might be too strict, maybe instead of failing the block it should just drop the trailing entries. Let me know what you think!

Reply:
@rob-solana can you share your thoughts here? What should we do if a slot doesn't end in a tick and so has trailing transaction entries?

Reply:
mark it as dead

Reply:
it'll sort of automatically die by dropping out of bank forks when we move root past it, so marking it dead is just an optimization

Reply:
thinking more about this: I'm not sure that blocktree exposes "this is the last entry in the transmission for the slot", so you might not be able to mark it dead until the root moves past

Reply:
After processing an entry slice, if we reached the bank's max tick height, we will freeze. So if we were to receive more entries, we wouldn't fetch them from the blocktree.

Reply:
I was speaking to the "trailing transaction entries" case.  Detecting such a case would be bubbling up the dataplane messages that indicate "end of slot", as it differs from max tick height.  I don't think we bubble that info up, so a slot with no final tick is currently just in limbo until we stop trying to complete it in replay_stage.

Reply:
Ah I see, I was referring to trailing entries after the last tick. Does your answer change at all for that scenario?

Reply:
yes: it'd be a good thing if entries after last tick marked the slot as dead.  the implication is that the node is hosed, possibly spamming, and the dataplane has no other defense against such a node (can exhaust disk).

Reply:
Great. I see, seems like blocktree can still get spammed with entries for a slot after replay stage "completes" a bank. Does that lead to any indeterminism? Are entry slices consistent for all nodes?

Reply:
resolved in call: yes.  if the last tick and the last shred message do no coincide there's a chance the network could be split on the block

replay_stage (via last tick) and blocktree (via last shred) should both think the slot is complete in order for the block to be accepted

Reply:
Determined that this is indeed the case, issue here: https://github.com/solana-labs/solana/issues/6498

Reply:
@aeyakovenko, @rob-solana, maybe marking this type of failure as a dead slot isn't the correct approach? Summary of concerns here: https://github.com/solana-labs/solana/issues/6498

Reply:
@carllin so from your last comment here: https://github.com/solana-labs/solana/issues/6498#issuecomment-545226727 is the conclusion that this PR is fine with the last entry check? Between https://github.com/solana-labs/solana/issues/6498 and https://github.com/solana-labs/solana/pull/6512 I think we're covered

Reply:
So from my understanding, in this PR, the situation that will trigger `BlockError::TrailingEntry` in both `replay_stage` and `blocktree_processor`:

Observed `max_tick_height`, and the last entry was not a tick (extra trailing entry after the last tick)

This is a slightly larger set of conditions under which `BlockError::TrailingEntry` should be triggered than what was proposed in #6512, which was:

Observed `max_tick_height`, and the last tick shred was not marked with `LAST_SHRED_IN_SLOT`.

Essentially the difference between the proposal and what's in this PR is, if the If the last tick shred was marked with `LAST_SHRED_IN_SLOT`, this slot shouldn't be marked as dead, even if there are trailing entries (because there might be a subset of this slot transmission that reaches consensus). If the last tick shred wasn't marked with `LAST_SHRED_IN_SLOT`,  then this should be marked as dead.

Reply:
@carllin the reason there's a separate issue for the "last tick needs to coincide with last shred" is that this PR doesn't cover it yet, and can land on its own.

Reply:
yup agreed, just summarizing the difference for clarity.